### PR TITLE
xDS implementation for enforcer

### LINKED ISSUE #1442: [Envoy-MG] Implement XDS APIs for the envoy gateway control plane
### Describe your problem(s)
The envoy proxy needed a control plane which implements the envoy xds APIs in order to dynamically configure it.

### Describe your solution
Implements the xds APIS as a separate service.

### How will you implement it
Envoy xds APIs implementation library written in Golang can be used for this purpose

### Tasks

- [x] [Basic Control Plane] Basic Control Plane implementation https://github.com/wso2/product-microgateway/issues/1466
- [x] [HTTPS Listener] HTTPS Listener Implementatin https://github.com/wso2/product-microgateway/issues/1465
- [x] [Refactor] Refactor Golang Implementation https://github.com/wso2/product-microgateway/issues/1468
- [x] [Routing] Route the requests based on HTTP Methods , resource level Endpoints https://github.com/wso2/product-microgateway/issues/1475
- [x] [HTTPS Upstreams] Support HTTPS Endpoints https://github.com/wso2/product-microgateway/issues/1477
- [x] [HTTPS Upstreams] Support Self-signed, Private CA based certificates and MTLS https://github.com/wso2/product-microgateway/issues/1496
- [x] [CORS] CORS Filter for the Proxy https://github.com/wso2/product-microgateway/issues/1494
- [x] [Sandbox/Production] Support Production Endpoints and Sandbox Endpoints separately https://github.com/wso2/product-microgateway/issues/1495
- [ ] [LB/ Failover] Loadbalance endpoints and Failover endpoints https://github.com/wso2/product-microgateway/issues/1497 
- [x] [HTTPS DownStreams] TLS communication for virtual hosts, listeners and MTLS 
---

### LINKED ISSUE #1445: [Envoy-MG] New configuration model and pass configurations from control plane to data plane
### Describe your problem(s)
Microgateway data plane has following two components

1. Envoy proxy
2. Filter chain

The control plane should be able to configure the both of the components.

### Describe your solution
Envoy proxy configuration can be done via the xds APIs.
Filter chain side configuration should be done using a similar approach used in envoy proxy.

### Tasks
- [x] Implement the filter chain config model #1452 
- [x] Dynamically update filter chain configs from control plane
---

### LINKED ISSUE #1452: Implement an configuration model for the filter chain
### Describe your problem(s)
In the filter chain side there should be a proper configuration model that can be read from file during the startup and initlaize the filter chain.

---

### PR DESCRIPTION
**Pull request by praminda**
### Purpose

This PR includes the initial xDS implementation of Enforcer Configuration and API discovery.
### Issues

Fixes #1445, Fixes #1452  

### Automation tests
 - Unit tests added: No
 - Integration tests added: No

### Tested environments

MacOS

---
#### Maintainers: Check before merge
- [x] Assigned 'Type' label
- [x] Assigned the project
- [x] Validated respective github issues
- [x] Assigned milestone to the github issue(s)

### CODE REVIEW

File: adapter/pkg/xds/xds_server.go [hunk 1]
```diff
@@ -274,3 +347,40 @@ func updateXdsCache(label string, endpoints []types.Resource, clusters []types.R
...
+
+	snap := cachev3.NewSnapshot(fmt.Sprint(version), nil, nil, nil, nil, nil, nil, configs, apis)
+	snap.Consistent()
+
+	err := enforcerCache.SetSnapshot(label, snap)
+	if err != nil {
+		logger.LoggerMgw.Error(err)
+	}
+
+	enforcerConfigVersionMap[label] = version
```
Comment (praminda) [inline adapter/pkg/xds/xds_server.go:371]: Read label from enforcer's env variable

File: enforcer/src/main/java/org/wso2/micro/gateway/enforcer/config/ConfigHolder.java [hunk 2]
```diff
@@ -67,42 +72,73 @@ public static ConfigHolder getInstance() {
      * Initialize the configuration provider class by reading the Mgw Configuration file.
      */
     private void init() {
-        String home = System.getenv(ConfigConstants.ENFORCER_HOME);
-        File file = new File(home + File.separator + ConfigConstants.CONF_DIR + File.separator + "config.toml");
-        configToml = new Toml().read(file).getTable(ConfigConstants.CONF_ENFORCER_TABLE);
-        config = configToml.to(EnforcerConfig.class);
-        readUnparsableConfigs();
+        String cdsHost = System.getenv().get(Constants.XDS_HOST);
```
Comment (praminda) [inline enforcer/src/main/java/org/wso2/micro/gateway/enforcer/config/ConfigHolder.java:75]: Add the same vars to k8s and other resources

File: enforcer/src/main/java/org/wso2/micro/gateway/enforcer/config/ConfigHolder.java [hunk 3]
```diff
@@ -67,42 +72,73 @@ public static ConfigHolder getInstance() {
...
+        authDto.setThreadPool(threadPool);
+
+        config.setAuthService(authDto);
+    }
+
+    private void populateEventHub(EventHub eventhub) {
+        EventHubConfigurationDto eventHubDto = new EventHubConfigurationDto();
+        eventHubDto.setEnable(eventhub.getEnabled());
+        eventHubDto.setServiceUrl(eventhub.getServiceUrl());
+//        eventHubDto.setJmsConnectionParameters(eventhub.getListenerEndpoint());
```
Comment (praminda) [inline enforcer/src/main/java/org/wso2/micro/gateway/enforcer/config/ConfigHolder.java:128]: Change to a properties bag

File: enforcer/src/main/java/org/wso2/micro/gateway/enforcer/config/ConfigHolder.java [hunk 4]
```diff
@@ -137,8 +173,8 @@ private void loadTrustStore() {
     }
 
     private void populateAPIMCredentials() {
-        String username = configToml.getString(ConfigConstants.APIM_CREDENTIAL_USERNAME);
-        String password = configToml.getString(ConfigConstants.APIM_CREDENTIAL_PASSWORD);
+        String username = "admin";
```
Comment (praminda) [inline enforcer/src/main/java/org/wso2/micro/gateway/enforcer/config/ConfigHolder.java:176]: Look at the xds dto adn set values

File: enforcer/src/main/java/org/wso2/micro/gateway/enforcer/constants/Constants.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+ * under the License.
+ */
+
+package org.wso2.micro.gateway.enforcer.constants;
+
+/**
+ * This class holds the constant keys related to the Microgateway.
+ */
+public class Constants {
+    public static final String XDS_HOST = "XDS_SERVER_HOST";
```
Comment (praminda) [inline enforcer/src/main/java/org/wso2/micro/gateway/enforcer/constants/Constants.java:25]: Change to `ADAPTER_HOST`

File: enforcer/src/main/java/org/wso2/micro/gateway/enforcer/discovery/ConfigDiscoveryClient.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+    public ConfigDiscoveryClient(String host, int port) {
+        channel = ManagedChannelBuilder.forAddress(host, port).usePlaintext().build();
+        this.blockingStub = ConfigDiscoveryServiceGrpc.newBlockingStub(channel);
+    }
+
+    public Config requestInitConfig() throws DiscoveryException {
+        // TODO: praminda implement nodeid (label) behavior
+        DiscoveryRequest req = DiscoveryRequest.newBuilder()
+                .setNode(Node.newBuilder().setId("enforcer").build())
```
Comment (praminda) [inline enforcer/src/main/java/org/wso2/micro/gateway/enforcer/discovery/ConfigDiscoveryClient.java:48]: get label from env var

File: enforcer/src/main/java/org/wso2/micro/gateway/enforcer/discovery/ConfigDiscoveryClient.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+                .setTypeUrl(Constants.CONFIG_TYPE_URL).build();
+        DiscoveryResponse res = DiscoveryResponse.getDefaultInstance();
+        try {
+            res = blockingStub.withDeadlineAfter(60, TimeUnit.SECONDS).fetchConfigs(req);
```
Comment (praminda) [inline enforcer/src/main/java/org/wso2/micro/gateway/enforcer/discovery/ConfigDiscoveryClient.java:52]: Add a retry

File: enforcer/src/main/java/org/wso2/micro/gateway/enforcer/discovery/ConfigDiscoveryClient.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+            shutdown();
+
+            // Theres only one config root resource here. Therefore taking 0, no need to iterate
+            return res.getResources(0).unpack(Config.class);
+        } catch (Exception e) {
+            // catching generic error here to wrap any grpc communication errors in the runtime
+            throw new DiscoveryException("Couldn't fetch init configs", e);
```
Comment (praminda) [inline enforcer/src/main/java/org/wso2/micro/gateway/enforcer/discovery/ConfigDiscoveryClient.java:59]: Throw a different exception type and exit from the top

**Comment by AmaliMatharaarachchi** [comment]
I suppose this PR does not fix all subtasks in the #1442 issue. If so, could you please update the pr description/issues accordingly?

**Comment by praminda** [comment]
Updated. Sorry I've linked the wrong issue number

**Review by AmaliMatharaarachchi — APPROVED** [review]

**Pull request merged**